packages/flutter/lib/src/material/search.dart

```diff
@@ -0,0 +1,416 @@
+/// used to exit the search page.
+///
+/// See also:
+///
+///  * [SearchDelegate] to define the content of the search page.
+Future<T> showSearch<T>({
+  @required BuildContext context,
+  @required SearchDelegate<T> delegate,
+  String query: '',
+}) {
+  assert(delegate != null);
+  assert(context != null);
+  assert(delegate._result == null || delegate._result.isCompleted);
+  delegate._result = new Completer<T>();
+  delegate.query = query ?? delegate.query;
+  delegate._currentBody.value = SearchBody.suggestions;
+  Navigator.of(context).push(new _SearchPageRoute<T>(
+    delegate: delegate,
+  ));
+  return delegate._result.future;
+}
+
+/// Delegate for [showSearch] to define the content of the search page.
+///
+/// The search page always shows an [AppBar] at the top where users can
+/// enter their search queries. The buttons shown before and after the search
+/// query text field can be customized via [SearchDelegate.leading] and
+/// [SearchDelegate.actions].
+///
+/// The body below the [AppBar] can either show suggested queries (returned by
+/// [SearchDelegate.buildSuggestions]) or - once the user submits a search  - the
+/// results of the search as returned by [SearchDelegate.buildResults].
+///
+/// [SearchDelegate.query] always contains the current query entered by the user
+/// and should be used to build the suggestions and results.
+///
+/// The results can be brought on screen by calling [SearchDelegate.showResults]
+/// and you can go back to showing the suggestions by calling
+/// [SearchDelegate.showSuggestions].
+///
+/// Once the user has selected a search result, [SearchDelegate.close] should be
+/// called to remove the search page from the top of the navigation stack and
+/// to notify the caller of [showSearch] about the selected search result.
+abstract class SearchDelegate<T> {
+
+  /// Suggestions shown in the body of the search page while the user types a
+  /// query into the search field.
+  ///
+  /// The delegate method is called whenever the content of [query] changes and
+  /// the value of [query] can be used to determine which suggestions should
+  /// be shown.
+  ///
+  /// Usually, this method will return a [ListView] with one [ListTile] per
+  /// suggestion. When [ListTile.onTap] is called, [query] should be updated
+  /// with the corresponding suggestion and the results page should be shown
+  /// by calling [showResults].
+  Widget buildSuggestions(BuildContext context);
+
+  /// The results shown after the user submits a search from the search page.
+  ///
+  /// The content of [query] can be used to determine what the users searched
+  /// for.
+  ///
+  /// Typically, this method returns a [ListView] with the search results.
+  Widget buildResults(BuildContext context);
+
+  /// A widget to display before the current query in the [AppBar].
+  ///
+  /// Typically an [IconButton] as a back button to exit the search. It is
```

Comment:
Typically an [IconButton] configured with a [BackButtonIcon] that exits the search with [close].

One can also use an [AnimatedIcon]...